server/server.nix

```diff
@@ -0,0 +1,92 @@
+{ aeson
+, async
+, base
+, bytestring
+, containers
+, directory
+, hashable
+, hspec
+, hspec-wai
+, http-types
+, jwt
+, mkDerivation
+, monad-logger
+, mtl
+, network
+, optparse-applicative
+, prometheus-client
+, prometheus-metrics-ghc
+, QuickCheck
+, quickcheck-instances
+, random
+, raven-haskell
+, scotty
+, sqlite-simple
+, securemem
+, stdenv
+, stm
+, text
+, time
+, unix
+, unordered-containers
+, uuid
+, wai
+, wai-extra
+, wai-middleware-prometheus
+, wai-websockets
+, warp
+, websockets
+}:
```

Comment:
What's the reason to have this list here not assigned to anything and then have the same packages listed again in `executableHaskellDepends`?

Reply:
Hey Nuno! If you're entirely new to Nix, I probably can't explain the entire thing from scratch here, but I'll try.

The quick summary is: this file defines a function which accepts an attribute set where we expect certain keys to be present.

```nix
# This is a function which accepts an attrset with keys 
# `foo` and `bar` and returns the `foo` key.
{ foo
, bar
}:
	foo
```

When we pass `./server.nix` to `callPackage` in `default.nix`, the `callPackage` function inspects the list of required attributes and wires these up with what it has in scope from Nixpkgs. This is a convention from Nixpkgs to express "I want this function to get the following arguments".

As you can see not every argument from the attrset goes into the same place. Some of the arguments go to `executableHaskellDepends`. Others go to `testHaskellDepends` or are used directly (`mkDerivation` or `stdenv.lib`)

Intro to the Nix language: https://nixos.org/nix/manual/#ch-expression-language

Packaging Haskell with Nix in https://github.com/Gabriel439/haskell-nix/

Reply:
Thank you so much, your explanation is clear and I get it now. Thanks also for the links, Nix can wait no more!